File: website/test/_lib/session.test.ts
import { describe, it, expect } from "vitest";
import {
  sealSession,
  openSession,
  buildSessionCookie,
  clearSessionCookie,
  readSessionCookie,
  defaultSessionExp,
  SESSION_COOKIE,
} from "../../functions/_lib/session";
import { makeEnv } from "./testutil";

const testEnv = makeEnv();

describe("web session cookie (AES-GCM)", () => {
  it("round-trips a sealed session", async () => {
    const sealed = await sealSession(testEnv, {
      sub: "user-1",
      display_name: "Alice",
      avatar_url: "https://img/a.png",
      exp: defaultSessionExp(),
    });
    const opened = await openSession(testEnv, sealed);
    expect(opened).not.toBeNull();
    expect(opened!.sub).toBe("user-1");
    expect(opened!.display_name).toBe("Alice");
    expect(opened!.avatar_url).toBe("https://img/a.png");
  });

  it("rejects a tampered ciphertext", async () => {
    const sealed = await sealSession(testEnv, {
      sub: "u",
      display_name: "d",
      avatar_url: "a",
      exp: defaultSessionExp(),
    });
    // flip a char in the ciphertext part
    const [iv, ct] = sealed.split(".");
    const tampered = `${iv}.${ct.slice(0, -1)}${ct.slice(-1) === "A" ? "B" : "A"}`;
    expect(await openSession(testEnv, tampered)).toBeNull();
  });

  it("rejects an expired session", async () => {
    const sealed = await sealSession(testEnv, {
      sub: "u",
      display_name: "d",
      avatar_url: "a",
      exp: Math.floor(Date.now() / 1000) - 10, // already expired
    });
    expect(await openSession(testEnv, sealed)).toBeNull();
  });

  it("rejects malformed cookie values", async () => {
    expect(await openSession(testEnv, "garbage")).toBeNull();
    expect(await openSession(testEnv, "a.b")).toBeNull();
    expect(await openSession(testEnv, "")).toBeNull();
  });

  it("cookie contains HttpOnly, Secure, SameSite=Lax", () => {
    const c = buildSessionCookie("value123");
    expect(c).toContain(`${SESSION_COOKIE}=value123`);
    expect(c).toContain("HttpOnly");
    expect(c).toContain("Secure");
    expect(c).toContain("SameSite=Lax");
    expect(c).toContain("Path=/");
  });

  it("clear cookie sets Max-Age=0", () => {
    expect(clearSessionCookie()).toContain("Max-Age=0");
  });

  it("reads the session value from a Cookie header", () => {
    const req = new Request("https://x/", {
      headers: { cookie: `other=1; ${SESSION_COOKIE}=abc.def; more=2` },
    });
    expect(readSessionCookie(req)).toBe("abc.def");
    expect(readSessionCookie(new Request("https://x/"))).toBeNull();
  });
});

File: server/internal/broker/session_test.go
package broker

import (
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

// TestSessionSealOpenRoundTrip mirrors session.test.ts round-trip.
func TestSessionSealOpenRoundTrip(t *testing.T) {
	sealer := newTestSealer(t)
	sealed, err := sealSession(sealer, webSession{
		Sub:         "user-1",
		DisplayName: "Alice",
		AvatarURL:   "https://img/a.png",
		Exp:         defaultSessionExp(),
	})
	if err != nil {
		t.Fatalf("seal: %v", err)
	}
	opened := openSession(sealer, sealed)
	if opened == nil {
		t.Fatal("opened session is nil")
	}
	if opened.Sub != "user-1" || opened.DisplayName != "Alice" || opened.AvatarURL != "https://img/a.png" {
		t.Errorf("opened = %+v", opened)
	}
}

// TestSessionRejectsTamper mirrors session.test.ts tamper rejection. It flips a
// base64url character in the INTERIOR of the ciphertext segment (not the final
// character): RawURLEncoding's last character only encodes the trailing 2/4 bits
// of a non-3-multiple length, so several distinct final characters decode to the
// same bytes — flipping only the last char can leave the ciphertext unchanged
// and would spuriously "open". An interior character maps to a full 6 bits, so
// changing it always alters the decoded ciphertext and must fail GCM auth.
func TestSessionRejectsTamper(t *testing.T) {
	sealer := newTestSealer(t)
	sealed, _ := sealSession(sealer, webSession{Sub: "u", DisplayName: "d", AvatarURL: "a", Exp: defaultSessionExp()})
	dot := strings.IndexByte(sealed, '.')
	if dot < 0 || dot >= len(sealed)-1 {
		t.Fatalf("unexpected sealed format: %q", sealed)
	}
	// Pick an interior index within the ciphertext segment (well before its
	// final character) so the mutation always changes real ciphertext bytes.
	idx := dot + 1
	orig := sealed[idx]
	repl := byte('A')
	if orig == 'A' {
		repl = 'B'
	}
	tampered := sealed[:idx] + string(repl) + sealed[idx+1:]
	if tampered == sealed {
		t.Fatal("tampered value should differ from the original")
	}
	if openSession(sealer, tampered) != nil {
		t.Error("tampered ciphertext should not open")
	}
}

// TestSessionRejectsExpired mirrors session.test.ts expiry rejection.
func TestSessionRejectsExpired(t *testing.T) {
	sealer := newTestSealer(t)
	sealed, _ := sealSession(sealer, webSession{Sub: "u", DisplayName: "d", AvatarURL: "a", Exp: nowSeconds() - 10})
	if openSession(sealer, sealed) != nil {
		t.Error("expired session should not open")
	}
}

// TestSessionRejectsMalformed mirrors session.test.ts malformed rejection.
func TestSessionRejectsMalformed(t *testing.T) {
	sealer := newTestSealer(t)
	for _, bad := range []string{"garbage", "a.b", ""} {
		if openSession(sealer, bad) != nil {
			t.Errorf("openSession(%q) should be nil", bad)
		}
	}
}

// TestSessionCookieAttributes verifies host-only cookie attributes (decision 1):
// HttpOnly + Secure + SameSite=Lax + Path=/ and NO Domain attribute.
func TestSessionCookieAttributes(t *testing.T) {
	c := buildSessionCookie("value123")
	got := c.String()
	for _, want := range []string{"lumen_session=value123", "HttpOnly", "Secure", "SameSite=Lax", "Path=/"} {
		if !strings.Contains(got, want) {
			t.Errorf("cookie %q missing %q", got, want)
		}
	}
	if strings.Contains(strings.ToLower(got), "domain=") {
		t.Errorf("session cookie must be host-only (no Domain): %q", got)
	}
}

// TestClearSessionCookie verifies the clear cookie sets Max-Age=0.
func TestClearSessionCookie(t *testing.T) {
	got := clearSessionCookie().String()
	if !strings.Contains(got, "Max-Age=0") {
		t.Errorf("clear cookie should set Max-Age=0: %q", got)
	}
}

// TestReadSessionCookie verifies reading the session value from a Cookie header.
func TestReadSessionCookie(t *testing.T) {
	req := httptest.NewRequest(http.MethodGet, "https://x/", nil)
	req.Header.Set("Cookie", "other=1; lumen_session=abc.def; more=2")
	if got := readSessionCookie(req); got != "abc.def" {
		t.Errorf("readSessionCookie = %q, want abc.def", got)
	}
	if got := readSessionCookie(httptest.NewRequest(http.MethodGet, "https://x/", nil)); got != "" {
		t.Errorf("readSessionCookie(no cookie) = %q, want empty", got)
	}
}

// TestAuthFlowRoundTrip verifies the auth-flow cookie seals verifier+state and
// rejects a payload missing them.
func TestAuthFlowRoundTrip(t *testing.T) {
	sealer := newTestSealer(t)
	sealed, err := sealAuthFlow(sealer, authFlowContext{Verifier: "v", State: "s", Exp: defaultAuthFlowExp()})
	if err != nil {
		t.Fatalf("seal: %v", err)
	}
	opened := openAuthFlow(sealer, sealed)
	if opened == nil || opened.Verifier != "v" || opened.State != "s" {
		t.Errorf("opened auth flow = %+v", opened)
	}

	// A session payload without verifier/state must not open as an auth flow.
	sessSealed, _ := sealSession(sealer, webSession{Sub: "u", Exp: defaultSessionExp()})
	if openAuthFlow(sealer, sessSealed) != nil {
		t.Error("session payload should not open as auth flow (missing verifier/state)")
	}
}
